File: Silver V/BOJ_17626/BOJ_17626.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int n;
vector<int> dp(50001, 2112345678);

void init()
{
    dp[0] = 0;
    for (int i = 1; i * i < 50001; ++i) {
        for (int j = 0; j < 50001; ++j) {
            if (j + i * i < 50001) {
                dp[j + i * i] = min(dp[j + i * i], dp[j] + 1);
            }
            else {
                break;
            }
        }
    }
}

int main()
{
    init();

    cin >> n;
    cout << dp[n] << '\n';

    return 0;
}

File: Silver V/BOJ_17626/BOJ_17626.py
import sys


def init(dp):
    dp[0] = 0
    # print(dp)
    for i in range(1, 250):
        start = i * i

        # print(start)
        for val in range(start, 50001, 1):
            # print(dp[val], dp[val - start])
            dp[val] = min(dp[val], dp[val - start] + 1)
            # print(dp[val])


if __name__ == "__main__":
    dp = [100000] * 50001
    init(dp)
    # print(dp)

    n = int(sys.stdin.readline())

    print(dp[n])
